src/coreComponents/meshUtilities/CPMesh/CPMeshBuilder.cpp

```diff
@@ -0,0 +1,409 @@
+              cPVertices( iCPVertex, 1 ),
+              cPVertices( iCPVertex, 2 ) );
+
+    // check if this vertex has already been found and inserted in the set
+    it = uniqueVerticesHelper.find( v );
+
+    // if already found, copy the local index in map and move on
+    if( it != uniqueVerticesHelper.end() )
+    {
+      Vertex const & existingVertex = *it;
+      localCPVertexToLocalVertex( iCPVertex ) = existingVertex.m_localVertex;
+    }
+    // if not found yet, insert into the set
+    else
+    {
+      v.m_localVertex = uniqueVerticesHelper.size();
+      localCPVertexToLocalVertex( iCPVertex ) = v.m_localVertex;
+      // if this vertex is owned by my rank,
+      // increment the counter (later used to assign global indices)
+      if( cPVertexIsOwned( iCPVertex ) )
+      {
+        nOwnedVertices++;
+      }
+      uniqueVerticesHelper.insert( v );
+    }
+  }
+
+  // Second step: move the data from the set to a array2d, because we don't want the set anymore
+  vertices.resizeDimension< 0, 1 >( uniqueVerticesHelper.size(), 3 );
+  localIndex & nLocalVertices = m_cPMeshData.nLocalVertices();
+  nLocalVertices = uniqueVerticesHelper.size();
+
+  std::for_each( uniqueVerticesHelper.begin(), uniqueVerticesHelper.end(), [&vertices]( Vertex const & v )
+  {
+    vertices( v.m_localVertex, 0 ) = v.m_x;
+    vertices( v.m_localVertex, 1 ) = v.m_y;
+    vertices( v.m_localVertex, 2 ) = v.m_z;
+  } );
+}
+
+void CPMeshBuilder::assignGlobalActiveCellIndices()
+{
+  // First step: gather all the nLocalActiveCells from all the ranks
+  localIndex const nOwnedActiveCells = m_cPMeshData.nOwnedActiveCells();
+  localIndex const offset = m_cPMeshComm.gatherLocalCountForEachRank( nOwnedActiveCells );
+
+  // Second step: using the offset for my rank, construct the local to global map
+  localIndex const nLocalActiveCells = m_cPMeshData.nLocalActiveCells();
+  array1d< bool > const & localCellIsOwned = m_cPMeshData.localCellIsOwned();
+  array1d< localIndex > const & localActiveCellToLocalCell = m_cPMeshData.localActiveCellToLocalCell();
+  array1d< globalIndex > & localActiveCellToGlobalActiveCell = m_cPMeshData.localActiveCellToGlobalActiveCell();
+  localActiveCellToGlobalActiveCell.resize( nOwnedActiveCells );
+
+  localIndex counter = 0;
+  for( localIndex iActiveCell = 0; iActiveCell < nLocalActiveCells; ++iActiveCell )
+  {
+    localIndex const iLocalCell = localActiveCellToLocalCell( iActiveCell );
+    if( localCellIsOwned( iLocalCell ) )
+    {
+      localActiveCellToGlobalActiveCell( iActiveCell ) = offset + counter;
+      counter++;
+    }
+    else
+    {
+      localActiveCellToGlobalActiveCell( iActiveCell ) = -1;
+    }
+  }
+
+  GEOSX_THROW_IF( counter != nOwnedActiveCells,
+                  "Partitioning error for the cells",
+                  InputError );
+
+  // for now, we don't care about global indices for the overlaps...
+}
+
+void CPMeshBuilder::assignGlobalVertexIndices()
```

Comment:
My initial plan was to use the position of the corner-point vertices to compute a global index that uniquely identifies each vertex in the mesh (something similar to what `nodeGlobalIndex` does in `InternalMeshGenerator`). However, corner-point meshes contain a lot of duplicate vertices. So, when I filter out duplicates, I remove a lot of vertices from the meshes (with their corresponding global indexes). 

As a result, in the final mesh, global indices are no longer contiguous. The largest global index is larger than the total number of vertices in the mesh, and some numbers in the interval [0, maxGlobalIndex] don't have any vertex associated with them. Hello @rrsettgast  @corbett5 @klevzoff do you know if having non-contiguous global indices can create problems somewhere in GEOSX? Maybe in `communicationTools` or in `dofManager`?

I also have another way to assign global vertex indices (MPI communications, I drafted something below), but it involves synchronization at the MPI partition boundaries and is much more complicated.

Reply:
AFAIK non-contiguous global indices aren't a problem. About the only thing I know that might not work is `LAIHelperFunctions::createPermutationMatrix`, which is just a debugging tool that has limitations.